source/simulator/solver_schemes.cc

```diff
@@ -573,19 +573,67 @@ namespace aspect
               {
                 const bool EisenstatWalkerChoiceOne = true;
                 parameters.linear_stokes_solver_tolerance = compute_Eisenstat_Walker_linear_tolerance(EisenstatWalkerChoiceOne,
-                                                            parameters.maximum_linear_stokes_solver_tolerance,
+                                                            newton_handler->get_maximum_linear_stokes_solver_tolerance(),
                                                             parameters.linear_stokes_solver_tolerance,
                                                             stokes_residual,
                                                             residual,
                                                             residual_old);
 
-                pcout << "   The linear solver tolerance is set to " << parameters.linear_stokes_solver_tolerance << std::endl;
+                const std::string preconditioner_stabilization = newton_handler->get_preconditioner_stabilization_string(newton_handler->get_preconditioner_stabilization());
+                const std::string velocity_block_stabilization = newton_handler->get_velocity_block_stabilization_string(newton_handler->get_velocity_block_stabilization());
+                pcout << "   The linear solver tolerance is set to " << parameters.linear_stokes_solver_tolerance << ". Stabilisation Preconditioner is " << preconditioner_stabilization << " and A block is " << velocity_block_stabilization << "." << std::endl;
               }
           }
 
         build_stokes_preconditioner();
 
-        stokes_residual = solve_stokes();
+        if (newton_handler->get_use_Newton_failsafe() == false)
+          {
+            stokes_residual = solve_stokes();
+          }
+        else
+          {
+            try
+              {
+                stokes_residual = solve_stokes();
+              }
+            catch (...)
+              {
+                computing_timer.exit_section();
```

Comment:
leave a comment here of the form
```
// start the solve over again and try with a stabilized version
```

Reply:
About line 602: you are exiting a timer block here, but what timer block does the remainder of this block correspond to? Don't you have to start a timer block again somewhere?

Reply:
So this exit solver.cc:525 `computing_timer.enter_section ("   Solve Stokes system");` (at the beginning of `solve_stokes()`. When the solver fails the `computing_timer.exit_section();` on line 921 is never executed, so I do it here.

Reply:
I don't think I understand this.

You have code like this:
```
  timer.enter_section (...);
  try {
    ...do something...
  } catch (...) {
    timer.leave_section();
    ...code now not associated with any timer section...
  }

  ...code that is now under the timer section again, but only if the try
     succeeded; if the try failed, this whole code block is not timed...
```
(I can't compare with the current code any more, because the line numbers you cite don't seem to match any more.)

Or are you saying that the `solve_stokes()` function enters a section at the top but the leaves through a `throw` statement with leaving the section? That would be a bug in that function -- it needs to leave the timer section before throwing the exception. If you don't right away see how to do this, show me the place tomorrow.

Reply:
>Or are you saying that the solve_stokes() function enters a section at the top but the leaves through a throw statement with leaving the section? That would be a bug in that function -- it needs to leave the timer section before throwing the exception. If you don't right away see how to do this, show me the place tomorrow.

Yes. I opened a separate pull request for this #2084 and removed it from this one.